### PR TITLE
[SYCL] Functions with variadic arguments should not get a diagnostic …

### PR DESCRIPTION
**Pull request by premanandrao**
…about calling convention

Signed-off-by: Premanand M Rao <premanand.m.rao@intel.com>

### CODE REVIEW

**Comment by premanandrao** [comment]
Thanks for the approval, Slava!

(I don't have merge permissions.   Can you or Alexey please do the merge?)

File: clang/lib/Sema/SemaType.cpp [hunk 1]
```diff
@@ -7024,13 +7024,12 @@ static bool handleFunctionTypeAttr(TypeProcessingState &state, ParsedAttr &attr,
...
 
       attr.setInvalid();
-      if (S.getLangOpts().SYCLIsDevice) {
-        S.SYCLDiagIfDeviceCode(attr.getLoc(), diag::err_cconv_varargs)
-          << FunctionType::getNameForCallConv(CC);
-        return true;
-      } else
-      return S.Diag(attr.getLoc(), diag::err_cconv_varargs)
-             << FunctionType::getNameForCallConv(CC);
+      // Disable this diagnostic for SYCL
```
Comment (bader) [inline clang/lib/Sema/SemaType.cpp]: SYCL 1.2.1 specification explicitly prohibits variadic functions:
"No variadic functions are allowed to be called in a SYCL kernel or any functions called by the kernel."

NOTE: variadic functions are allowed in the "host code".

File: clang/test/SemaSYCL/sycl-varargs-cconv.cpp [hunk 2]
```diff
@@ -0,0 +1,4 @@
+// RUN: %clang_cc1 -fsycl-is-device -verify -fsyntax-only -x c++ -emit-llvm-only %s
+// expected-no-diagnostics
+
+int __cdecl foo(int, ...);
```
Comment (bader) [inline clang/test/SemaSYCL/sycl-varargs-cconv.cpp]: Please, add test case with the `foo` call from the SYCL kernel and check that compiler emits an error.

File: clang/lib/Sema/SemaType.cpp (hunk 1 above)
Comment (premanandrao) [reply clang/lib/Sema/SemaType.cpp]: Hi Alexey,

I want to issue this diagnostic at the call site - i.e. when such a function is called in the kernel context.   Currently the diagnostic is issued when the declaration is simply seen.   That seems wrong and that is what I have disabled.

File: clang/lib/Sema/SemaType.cpp [hunk 3]
```diff
@@ -7024,13 +7024,12 @@ static bool handleFunctionTypeAttr(TypeProcessingState &state, ParsedAttr &attr,
...
+      // Note that deferring does not work since the declaration
+      // need not be in a function context.
```
Comment (bader) [inline clang/lib/Sema/SemaType.cpp]: I guess SYCL Compiler should skip the diagnostic for declaration, but emit an error if vararg function is called. Right?

File: clang/test/SemaSYCL/sycl-varargs-cconv.cpp (hunk 2 above)
Comment (premanandrao) [reply clang/test/SemaSYCL/sycl-varargs-cconv.cpp]: May I do that as a separate patch or do you want that along with this one?

File: clang/lib/Sema/SemaType.cpp (hunk 3 above)
Comment (premanandrao) [reply clang/lib/Sema/SemaType.cpp]: Yes.  I wanted to do the latter as a separate patch, as currently this is preventing the compilation of common header files.

**Comment by mibintc** [comment]
Also, Else after return is against llvm coding conventions

**Comment by premanandrao** [comment]
> Also, Else after return is against llvm coding conventions

There is no "else after return" in the code I added new.

**Comment by mibintc** [comment]
If you are going to upload another patch will you please correct the else after return, that will save some work in the future.

**Comment by premanandrao** [comment]
> If you are going to upload another patch will you please correct the else after return, that will save some work in the future.

I am not sure which "else after return" you are talking about.   If you are talking about the part I removed, that will stay removed.   My new changes will most likely be an addition to another file.  (i.e. this particular change that I uploaded will remain unchanged.)

File: clang/test/SemaSYCL/sycl-varargs-cconv.cpp [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  bar(7.0f); // expected-no-error
+}
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
+  kernelFunc();
+}
+
+int main() {
+  //expected-error@+1 {{SYCL kernel cannot call a variadic function}}
```
Comment (bader) [inline clang/test/SemaSYCL/sycl-varargs-cconv.cpp]: I think, we need to call `bar` from the host code to make sure that compiler do not emit errors.

File: clang/lib/Sema/SemaType.cpp [hunk 5]
```diff
@@ -7024,13 +7024,12 @@ static bool handleFunctionTypeAttr(TypeProcessingState &state, ParsedAttr &attr,
...
+      if (!S.getLangOpts().SYCLIsDevice)
```
Comment (bader) [inline clang/lib/Sema/SemaType.cpp]: I think we just need to remove all SYCL specific code from this part as varargs function is detected by SemaSYCL checks.
Could you clarify why this check is still needed? 
Is the problem is `spir_function` calling convention or `__cdecl`?

File: clang/test/SemaSYCL/sycl-include-gnu11.cpp [hunk 6]
```diff
@@ -0,0 +1,22 @@
+// RUN: %clang_cc1 -triple spir64-unknown-linux-sycldevice -fcxx-exceptions -fsycl-is-device -verify -fsyntax-only -std=gnu++11 %s -I %S/Inputs
+
+// This test verifies that sycl.hpp can be compiled cleanly with -std=gnu++11
```
Comment (bader) [inline clang/test/SemaSYCL/sycl-include-gnu11.cpp]: I would prefer this to committed in a separate commit as this test has noting to do with this PR.

Also this test should be full stack - please, move it to sycl/test and add execution of the compiled app.
This version do not cover the driver functionality.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 7]
```diff
@@ -353,7 +353,12 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
           return false;
         }
       }
+    } else if (const auto *ATy = dyn_cast<AttributedType>(Ty)) {
+      return CheckSYCLType(ATy->getModifiedType(), Loc, Visited);
```
[the remark this answers is not in the document]
Comment (premanandrao) [reply clang/lib/Sema/SemaSYCL.cpp]: Done, thanks.

File: clang/test/SemaSYCL/sycl-varargs-cconv.cpp (hunk 4 above)
Comment (premanandrao) [reply clang/test/SemaSYCL/sycl-varargs-cconv.cpp]: I added a call.

File: clang/lib/Sema/SemaType.cpp (hunk 5 above)
Comment (premanandrao) [reply clang/lib/Sema/SemaType.cpp]: You are right, this is no longer needed.   I have remove the SYCL-specific changes.

File: clang/test/SemaSYCL/sycl-include-gnu11.cpp (hunk 6 above)
Comment (premanandrao) [reply clang/test/SemaSYCL/sycl-include-gnu11.cpp]: I have removed this test.  I will add it as a separate change.

**Pull request merged**